src/runtime/uv/timer.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Sofia Rodrigues
+*/
+#include "runtime/uv/timer.h"
+
+namespace lean {
+#ifndef LEAN_EMSCRIPTEN
+
+using namespace std;
+
+// The finalizer of the `Timer`.
+// TODO: check if it can cause the promise to leak in some way.
+void lean_uv_timer_finalizer(void* ptr) {
+    lean_uv_timer_object * timer_obj = (lean_uv_timer_object*) ptr;
+    timer_obj->m_uv_timer.data = ptr;
+
+    if (timer_obj->m_promise != NULL) lean_dec(timer_obj->m_promise);
+
+    event_loop_lock(&global_ev);
+
+    uv_close((uv_handle_t*) & timer_obj->m_uv_timer, [](uv_handle_t* handle) {
+        free((lean_uv_timer_object*)handle->data);
+    });
+
+    event_loop_unlock(&global_ev);
+}
+
+void initialize_libuv_timer() {
+    g_uv_timer_external_class = lean_register_external_class(lean_uv_timer_finalizer, [](auto /**/, auto /**/) {});
+}
+
+void handle_timer_event(uv_timer_t* handle) {
+    lean_object * obj = (lean_object*)handle->data;
+    lean_uv_timer_object * timer = lean_to_uv_timer(obj);
+
+    // If the last promise is already solved then we can just set it as closed.
+    // and free the object.
+    if(!timer->m_repeating || (timer->m_promise != NULL && lean_io_get_task_state_core(timer->m_promise) == 2)) {
```

Comment:
But right now with `reset` we only reset the timer, we cannot stop them, I'm not sure if it's the right behavior because if the event loop owns the timer, and we will not use the timer again, the memory will leak and the timer will run forever.

Reply:
Right, I spent some time thinking about this and the fundamental issue here is that we have both `IO.Promise` and `Timer` as separately ref counted objects. The fundamental invariant that we need to uphold is the following:
1. While there is an unresolved `Promise` associated with the timer we need to keep the `Timer` object alive so it can eventually resolve the promise.
2. If there is no unresolved `Promise` associated with the timer we should drop it ASAP.

Given the fact that we have both separate objects and this constraint the proper solution for upholding this invariant to me seems to be:
1. Upon creation of `Timer` the ref count is just 1
2. Upon `Timer.next` we increment the ref count of both the `Timer` and the `Promise` to 2. This makes sure that if either the `Timer` or the `Promise` ends up being dropped by the user, the callback still gets valid memory passed to it.
3. When the execution of the callback is finished we decrement the ref counter of the `Timer` and the associated `Promise`.

This ensure that even when the `Timer` is dropped by user code before the `Promise` is resolved the `Promise` is still going to be resolved and then the timer will be freed. If the user drops both the `Timer` and the `Promise` before the tick has finished (seems unlikely but hey) the memory will live until the first tick of the timer has finished and then be freed.

I do not think that we can do better than this while the notion of a `Promise` and a `Timer` are separated. Rust manages to solve this correctly as the resources of a `Future` are part of the object that implements the `Future` trait so resources and the awaitable thing are stored by the same struct that gets freed at the same time.

Additionally we can add a function `Timer.stop` that forcibly stops the timer, dropping the associated `Promise` and putting it back into the initial state. However users need of course be mindful about using this feature as it might happen, that they end up with a leaked task in this fashion.

Reply:
Implemented this behavior :D